### ISSUE TITLE
Dates being truncated and not ISO-8601 format

### ISSUE THREAD

**Issue by thetylerhayes** [blue-button/bluebutton.js#123]
This is a vital's date taken from a real-world C32 file: `<effectiveTime value="20141230232334" />`

However in the JSON `vitals` section, it's represented as: `"date": "12/30/2014"`

Problems:
1. No time, just date
2. Not ISO standard like `"date": "2014-12-30T23:23:34.000Z",`. I.e., it's MM/DD/YYYY instead of YYYY-MM-DD.

**Comment by thetylerhayes** [comment]
To be clear, I'm not saying the dates should be represented in JSON the same as they are in the XML file. There just needs to be one consistent date format used in the JSON.

**Comment by molsches** [comment]
I'd vote for ISO-8601 timestamps consistently (if there are votes)

**Comment by thetylerhayes** [comment]
Same but I just want people to like me.

**Comment by owenfi** [comment]
vitalsTable might be Prime internal :) ?

**Comment by thetylerhayes** [comment]
Ha yeah fuck I totally showed our hand there.

Well either way dates are being truncated. Not good. Also unclear why we changed from ISO-8601 to Betamax.

**Comment by sankarravi** [comment]
See https://github.com/elationemr/bluebutton.js/commit/65236a2d37efaf42453c8409afc71283eb8549dd for the sad JS datetime parsing that led me to Betamax :). The short version:
```
* We output strings like 04/27/2014 because "2014-04-27" will be parsed
* assuming midnight UTC (and so will turn to something like 04/26/2014 at 5PM).
```
Let me know if you have other ideas about how to solve.

I don't know why `vitals` is losing precision in the original document, though. That's still worth looking into.

**Comment by thetylerhayes** [comment]
@sankarravi Does that not also happen on MM/DD/YYYY dates? If not, why not?

Seems strange I've never heard of this being an issue before. Sounds like a big deal.

**Comment by sankarravi** [comment]
Nope. "MM/DD/YYYY" seems to get interpreted as "midnight local time" and "YYYY-MM-DD" as "midnight UTC."
Chrome:
```
> d1 = new Date("01/05/2015")
Mon Jan 05 2015 00:00:00 GMT-0800 (PST)
> d2 = new Date("2015-01-05")
Sun Jan 04 2015 16:00:00 GMT-0800 (PST)
```
Node:
```
> d1 = new Date("01/05/2015")
Mon Jan 05 2015 00:00:00 GMT-0800 (PST)
> d2 = new Date("2015-01-05")
Sun Jan 04 2015 16:00:00 GMT-0800 (PST)
```
I have no idea why that's not all over the internet...

But in any case, given that situation, I think the MM/DD/YYYY format ends up requiring the least diligence for library users, when we don't have time information. If we wanted to consistently use that format, I think we could write stuff like "12/30/2014 23:23:34Z" and it should parse equivalently to the ISO string (because the Z specifies UTC). I kind of like using the ISO format when it doesn't break shit, but a case for consistency seems reasonable too.

**Comment by thetylerhayes** [comment]
This has some great, deep breakdown: http://stackoverflow.com/questions/2587345/javascript-date-parse

My personal favorite answer is the super concise one: 

> There is some method to the madness. As a general rule, if a browser can interpret a date as an ISO-8601, it will. "2005-07-08" falls into this camp, and so it is parsed as UTC. "Jul 8, 2005" cannot, and so it is parsed in the local time.

> See [JavaScript and Dates, What a Mess!](http://blog.dygraphs.com/2012/03/javascript-and-dates-what-mess.html) for more.

**Comment by thetylerhayes** [comment]
And a whole lot of pragmatic recommendations here: http://blog.dygraphs.com/2012/03/javascript-and-dates-what-mess.html

> Here are some rules of thumb:

> Stick to "YYYY/MM/DD" for your date strings whenever possible. It's universally supported and unambiguous. With this format, all times are local.
Avoid using hyphenated dates ("YYYY-MM-DD") unless you know what you're doing. Only newer browsers support them.
So long as you specify four digits years, date strings at least have a unique parse across all browsers. Some browsers may return an error, but you won't get inconsistent behavior.
Chrome tends to be more accepting than other browsers. If a date format parses in Chrome, you shouldn't assume that it parses anywhere else.
If a recent browser can interpret as date string as ISO-8601, it will. With this format, your date/time string is interpreted as UTC.

**Comment by thetylerhayes** [comment]
All of which is to say: good work @sankarravi :+1: 

Now just need to figure out why the dates are losing precision ;)

**Comment by sankarravi** [comment]
Ok, well the issue turns out not to be very mysterious. This is our parseDate function:
https://github.com/blue-button/bluebutton.js/blob/master/lib/documents.js#L53

It makes literally no effort to parse anything except year/month/day :)

So then the question is how to fix it. I started to augment our date parsing function to look at times. Something like this:
```
if (str.length >= 14) {
  var hour = str.substr(8, 2);
  var min = str.substr(10, 2);
  var secs = str.substr(12, 2);

  return new Date(Date.UTC(year, month, day, hour, min, secs));
}
```
But then I noticed that some of the files in our test suites had dates like `20050329224411+0500`. That seemed like a pain to a parse for, especially with the possibility of a colon in the middle or milliseconds before the timezone... And also we have some dates like "199911" that we are currently parsing incorrectly (that one turns into "10/31/1999"!). So then I went looking for something someone else has already written.

I couldn't really find any options besides moment.js. Which seems super well loved and battle tested, but it means pulling in a new dependency (I had been hoping to find a 200L thing somewhere I could just copy in). it's 35KB minified... compared to 50KB for BB.js itself. I only use BB.js from Node, so other than preserving the ability to have a pretty sandbox demo on bluebuttonjs.com, the browser-side doesn't bother me very much... Does adding a new dependency bother anyone else? I know @blacktm has historically attempted to keep out external dependencies for the browser's sake, but I do feel like date handling is both important and notoriously tricky.

One option I've considered is including fallback code like we have for EJS, where we keep the old/naive date-parsing around for people who really don't want to add moment.js in the browser and rely on moment if it exists. It's less obvious to me that that is a good solution here than it was there, though, because it seems weird that the library produces less correct data if you don't have a dependency that we make optional. It might be better to just break.

So, those following this issue: opinions?
(I do have the start of a branch using moment.js and it seems like it clearly gets more accurate data out. The remaining task is just for me to fix all the tests by manually verifying that new dates are what we want them to be based on what's in the XML)

**Comment by thetylerhayes** [comment]
https://github.com/rvanbaalen/date-steroids is 5 KB?

**Comment by sankarravi** [comment]
Looks like that just does date formatting (taking an existing Date object and turning it into a string); what we want is the opposite (taking the string and parsing it into a Date object). Maybe I'll look at whether I can extract that out of moment.js and end up with something much smaller than the entire library. Let me know if you find anything else that appears to be doing that already!

**Comment by sankarravi** [comment]
Made an effort to extract the parsing code out of Moment, but it was pretty complicated, and I ended up needing to pull lots of the library along. Going to try to make a second effort where I use the moment branch to correct the expected output in all of the JSON and then try to augment our date-parsing to do produce the same results... we'll see where I get there.

**Comment by sankarravi** [comment]
Released 0.3.0
